src/query/processor.rs

```diff
@@ -122,14 +160,142 @@ impl<T: Transport + Clone> Processor<T> {
+        }
+    }
+
     pub fn status(&self, query_id: QueryId) -> Option<QueryStatus> {
         self.queries.handle(query_id).status()
     }
+
+    /// Handle the next command from the input stream.
+    pub async fn handle_next(&mut self) {
+        if let Some(command) = self.command_stream.next().await {
+            match command.payload {
+                TransportCommand::Query(query_command) => match query_command {
+                    QueryCommand::Create(req, resp) => {
+                        let result = self.new_query(req).await.unwrap();
+                        resp.send(result).unwrap();
+                    }
+                    QueryCommand::Prepare(req) => {
+                        self.prepare(req).await.unwrap();
+                    }
+                    QueryCommand::Input(query_input) => {
+                        self.receive_inputs(query_input).unwrap();
+                    }
+                },
+                TransportCommand::StepData(_, _, _) => panic!("unexpected command: {command:?}"),
+            }
+        }
+    }
+
+    /// Awaits the query completion
+    pub async fn complete(
+        &mut self,
+        query_id: QueryId,
+    ) -> Result<Box<dyn ProtocolResult>, QueryCompletionError> {
+        let handle = {
+            let mut queries = self.queries.inner.lock().unwrap();
+
+            match queries.remove(&query_id) {
+                Some(QueryState::Running(handle)) => {
+                    queries.insert(query_id, QueryState::AwaitingCompletion);
+                    Ok(handle)
+                }
+                Some(state) => {
+                    let state_error = StateError::InvalidState {
+                        from: QueryStatus::from(&state),
+                        to: QueryStatus::Running,
+                    };
+                    queries.insert(query_id, state);
+                    Err(QueryCompletionError::StateError {
+                        source: state_error,
+                    })
+                }
+                None => Err(QueryCompletionError::NoSuchQuery(query_id)),
+            }
+            //
+            // if let Some(QueryState::Running(handle)) = queries.remove(&query_id) {
+            //     queries.insert(query_id, QueryState::AwaitingCompletion);
+            //     Ok(handle)
+            // } else {
+            //     queries.insert(query_id, );
+            //     Err(QueryCompletionError::StateError {
+            //         source: StateError::InvalidState {
+            //             from: QueryStatus::from(&state),
+            //             to: QueryStatus::Running,
+            //         },
+            //     })
+            // }
+        }?;
+
+        Ok(handle.await.unwrap())
+
+        // match queries.entry(query_id) {
+        //     Entry::Occupied(entry) => {
+        //         let state = entry.remove();
+        //         drop(queries);
```

Comment:
fwiw i've tried doing this kind of thing before, and rust cannot understand that `drop` means you aren't carrying the (not-`Send`) mutex guard across an `await`. i've found the only thing that works is putting braces around it. something to the effect of

```rust
let state: Option<QueryState> = {
    let mut queries = self.queries.inner.lock().unwrap();
    queries.remove(query_id)
};
```
maybe moot point if you aren't going to use this code, but yeah

Reply:
true, I had to do the same trick